File: Java/labs/lab15_Command/task_2_1 – solution/src/TextEditor.java
public class TextEditor {

    private Document document;

    final private Button saveButton;
    final private Button printButton;

    final private MenuItem saveMenuItem;
    final private MenuItem printMenuItem;
    final private MenuItem showInfoMenuItem;

    final private Shortcut saveShortcut;
    final private Shortcut printShortcut;

    public TextEditor() {
        saveButton = new Button();
        saveButton.setCommand(saveDocument);
        printButton = new Button();
        printButton.setCommand(printDocument);

        saveMenuItem = new MenuItem();
        saveMenuItem.setCommand(saveDocument);
        printMenuItem = new MenuItem();
        printMenuItem.setCommand(printDocument);
        showInfoMenuItem = new MenuItem();
        showInfoMenuItem.setCommand(showInfo);

        saveShortcut = new Shortcut();
        saveShortcut.setCommand(saveDocument);
        printShortcut = new Shortcut();
        printShortcut.setCommand(printDocument);
    }

    final private Command saveDocument = new Command() {
        @Override
        public void execute() {
            if (document != null){
                document.save();
            }
        }
    };

    final private Command printDocument = new Command() {
        @Override
        public void execute() {
            if (document != null){
                document.print();
            }
        }
    };

    final private Command showInfo = new Command() {
        @Override
        public void execute() {
            if (document != null){
                document.showInfo();
            }
        }
    };

    public void newDocument(String name) {
        document = new Document(name);
    }

    public void clickSaveButton() {
        saveButton.click();
    }

    public void clickPrintButton() {
        printButton.click();
    }

    public void selectSaveMenuItem() {
        saveMenuItem.select();
    }

    public void selectPrintMenuItem() {
        printMenuItem.select();
    }

    public void selectShowInfoMenuItem() {
        showInfoMenuItem.select();
    }

    public void pressSaveShortcut() {
        saveShortcut.press();
    }

    public void pressPrintShortcut() {
        printShortcut.press();
    }
}

File: Cpp/lab15_Command/task_2_1 - solution/src/text_editor.cpp
#include "text_editor.hpp"

#include <iostream>


TextEditor::SaveDocument::SaveDocument(Document*& document)
: document(document) {}

TextEditor::SaveDocument::~SaveDocument() {}

void TextEditor::SaveDocument::execute()
{
  if (this->document != nullptr)
  {
    this->document->save();
  }
}

TextEditor::PrintDocument::PrintDocument(Document*& document)
: document(document) {}

TextEditor::PrintDocument::~PrintDocument() {}

void TextEditor::PrintDocument::execute()
{
  if (this->document != nullptr)
  {
    this->document->print();
  }
}

TextEditor::ShowInfo::ShowInfo(Document*& document)
: document(document) {}

TextEditor::ShowInfo::~ShowInfo() {}

void TextEditor::ShowInfo::execute()
{
  if (this->document != nullptr)
  {
    this->document->showInfo();
  }
}


TextEditor::TextEditor() :
  saveButton(Button()),
  printButton(Button()),

  saveMenuItem(MenuItem()),
  printMenuItem(MenuItem()),
  showInfoMenuItem(MenuItem()),

  saveShortcut(Shortcut()),
  printShortcut(Shortcut())
{
  this->saveButton.setCommand(
    new TextEditor::SaveDocument(this->document)
  );
  this->printButton.setCommand(
    new TextEditor::PrintDocument(this->document)
  );

  this->saveMenuItem.setCommand(
    new TextEditor::SaveDocument(this->document)
  );
  this->printMenuItem.setCommand(
    new TextEditor::PrintDocument(this->document)
  );
  this->showInfoMenuItem.setCommand(
    new TextEditor::ShowInfo(this->document)
  );

  this->saveShortcut.setCommand(
    new TextEditor::SaveDocument(this->document)
  );
  this->printShortcut.setCommand(
    new TextEditor::PrintDocument(this->document)
  );
}

TextEditor::~TextEditor() {}

void TextEditor::newDocument(std::string const& name)
{
  this->document = new Document(name);
}

void TextEditor::clickSaveButton()
{
  this->saveButton.click();
}

void TextEditor::clickPrintButton()
{
  this->printButton.click();
}

void TextEditor::selectPrintMenuItem()
{
  this->printMenuItem.select();
}

void TextEditor::selectSaveMenuItem()
{
  this->saveMenuItem.select();
}

void TextEditor::selectShowInfoMenuItem()
{
  this->showInfoMenuItem.select();
}

void TextEditor::pressSaveShortcut()
{
  this->saveShortcut.press();
}

void TextEditor::pressPrintShortcut()
{
  this->printShortcut.press();
}
